### PR TITLE
Add ElasticSearch integration

### PR DESCRIPTION
**Pull request by sharat87**

### CODE REVIEW

**Comment by sharat87** [comment]
Implementation for #419.

File: app/server/appsmith-plugins/elasticSearchPlugin/src/main/java/com/external/plugins/ElasticSearchPlugin.java [hunk 1]
```diff
@@ -0,0 +1,189 @@
...
+            String body = actionConfiguration.getBody();
+
+            final String path = actionConfiguration.getPath();
+            final Request request = new Request(actionConfiguration.getHttpMethod().toString(), path);
+            ContentType contentType = ContentType.APPLICATION_JSON;
+
+            if (isBulkQuery(path)) {
+                contentType = ContentType.create("application/x-ndjson");
+
+                // If body is a JSON Array, instead of ND-JSON, convert it to ND-JSON.
```
Comment (trishaanand) [inline app/server/appsmith-plugins/elasticSearchPlugin/src/main/java/com/external/plugins/ElasticSearchPlugin.java:63]: I think the comment needs rewording. It says that instead of A, covert it to A.

File: app/server/appsmith-plugins/elasticSearchPlugin/src/main/java/com/external/plugins/ElasticSearchPlugin.java [hunk 2]
```diff
@@ -0,0 +1,189 @@
...
+                if (body.trim().startsWith("[")) {
+                    final StringBuilder ndJsonBuilder = new StringBuilder();
+                    try {
+                        List<Object> commands = objectMapper.readValue(body, ArrayList.class);
+                        for (Object object : commands) {
+                            ndJsonBuilder.append(objectMapper.writeValueAsString(object)).append("\n");
+                        }
+                    } catch (IOException e) {
+                        return Mono.error(e);
```
Comment (trishaanand) [inline app/server/appsmith-plugins/elasticSearchPlugin/src/main/java/com/external/plugins/ElasticSearchPlugin.java:72]: Could we throw a Plugin Exception here instead of generic Mono.error?

File: app/server/appsmith-plugins/elasticSearchPlugin/src/main/java/com/external/plugins/ElasticSearchPlugin.java [hunk 3]
```diff
@@ -0,0 +1,189 @@
...
+
+            final String path = actionConfiguration.getPath();
+            final Request request = new Request(actionConfiguration.getHttpMethod().toString(), path);
+            ContentType contentType = ContentType.APPLICATION_JSON;
+
+            if (isBulkQuery(path)) {
+                contentType = ContentType.create("application/x-ndjson");
+
+                // If body is a JSON Array, instead of ND-JSON, convert it to ND-JSON.
+                if (body.trim().startsWith("[")) {
```
Comment (trishaanand) [inline app/server/appsmith-plugins/elasticSearchPlugin/src/main/java/com/external/plugins/ElasticSearchPlugin.java:64]:
```suggestion
                if (body != null && body.trim().startsWith("[")) {
```

File: app/server/appsmith-plugins/elasticSearchPlugin/src/main/java/com/external/plugins/ElasticSearchPlugin.java [hunk 4]
```diff
@@ -0,0 +1,189 @@
...
+            if (body != null) {
+                request.setEntity(new NStringEntity(body, contentType));
+            }
+
+            try {
+                final String responseBody = new String(
+                        client.performRequest(request).getEntity().getContent().readAllBytes());
+                result.setBody(objectMapper.readValue(responseBody, HashMap.class));
+            } catch (IOException e) {
+                return Mono.error(e);
```
Comment (trishaanand) [inline app/server/appsmith-plugins/elasticSearchPlugin/src/main/java/com/external/plugins/ElasticSearchPlugin.java:87]: Similar request. Could we use wrap the exception inside Appsmith's Plugin Exception? Thanks!

File: app/server/appsmith-plugins/elasticSearchPlugin/src/main/java/com/external/plugins/ElasticSearchPlugin.java [hunk 5]
```diff
@@ -0,0 +1,189 @@
...
+            try {
+                client.close();
+            } catch (IOException e) {
+                log.warn("Error closing connection to ElasticSearch.", e);
+            }
+        }
+
+        @Override
+        public Set<String> validateDatasource(DatasourceConfiguration datasourceConfiguration) {
+            return null;
```
Comment (trishaanand) [inline app/server/appsmith-plugins/elasticSearchPlugin/src/main/java/com/external/plugins/ElasticSearchPlugin.java:148]: Is this not required? Are there no mandatory fields that need to validated? I am sorry. I haven't looked much into Elastic Search and am coming from an ignorant state.

File: app/server/appsmith-plugins/elasticSearchPlugin/src/main/java/com/external/plugins/ElasticSearchPlugin.java (hunk 1 above)
Comment (sharat87) [reply app/server/appsmith-plugins/elasticSearchPlugin/src/main/java/com/external/plugins/ElasticSearchPlugin.java:63]: Haha, yeah. Let me try again.

File: app/server/appsmith-plugins/elasticSearchPlugin/src/main/java/com/external/plugins/ElasticSearchPlugin.java (hunk 2 above)
Comment (sharat87) [reply app/server/appsmith-plugins/elasticSearchPlugin/src/main/java/com/external/plugins/ElasticSearchPlugin.java:72]: Done.

File: app/server/appsmith-plugins/elasticSearchPlugin/src/main/java/com/external/plugins/ElasticSearchPlugin.java (hunk 4 above)
Comment (sharat87) [reply app/server/appsmith-plugins/elasticSearchPlugin/src/main/java/com/external/plugins/ElasticSearchPlugin.java:87]: Done.

File: app/server/appsmith-plugins/elasticSearchPlugin/src/main/java/com/external/plugins/ElasticSearchPlugin.java (hunk 5 above)
Comment (sharat87) [reply app/server/appsmith-plugins/elasticSearchPlugin/src/main/java/com/external/plugins/ElasticSearchPlugin.java:148]: Implemented validation, sorry, missed it previously.

**Review by trishaanand — APPROVED** [review]

**Review by 641210 — COMMENTED** [review]
[@image]()
